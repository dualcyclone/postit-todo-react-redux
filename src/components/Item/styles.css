li {
    padding: 5px;
    margin: 1em;
    float: left;
    list-style: none;
}

li div {
    text-decoration: none;
    font-family: 'Nothing You Could Do', cursive;
    color: #000;
    display: block;
    height: 10em;
    width: 10em;
    padding: 1em;
    box-shadow: 5px 5px 7px rgba(33,33,33,.7);
    transition: -webkit-transform .15s linear;
    transform: rotate(-6deg);
    overflow: hidden;
}

li.complete div p {
    text-decoration: line-through;
}

li.complete div p:before {
    position: absolute;
    content: '';
    background: #000;
    display: block;
    width: 100%;
    height: 1px;
    transform: rotate(-50deg);
    left: 0;
    right: 0;
    top: 0;
    bottom: 0;
    margin: auto;
}

li.complete div p:after {
    position: absolute;
    content: '';
    background: #000;
    display: block;
    width: 100%;
    height: 1px;
    transform: rotate(50deg);
    left: 0;
    right: 0;
    top: 0;
    bottom: 0;
    margin: auto;
}

li:nth-child(even) div {
    transform: rotate(4deg);
    position: relative;
    top: 5px;
}

li:nth-child(3n) div {
    transform: rotate(-3deg);
    position: relative;
    top: -5px;
}

li:nth-child(4n) div {
    transform: rotate(2deg);
    position: relative;
    top: -5px;
}

li:nth-child(5n) div {
    transform: rotate(5deg);
    position: relative;
    top: -10px;
}

li div:hover, li div:focus{
    box-shadow: 10px 10px 7px rgba(0,0,0,.7);
    transform: scale(1.25);
    position: relative;
    z-index: 5;
}

li div .icon {
    display: block;
    position: absolute;
    visibility: hidden;
}

li div:hover .icon {
    visibility: visible;
}

li div .icon {
    color: #000;
}

li div .icono-trash {
    bottom: 2px;
    right: 2px;
}

li div .icono-check, li div .icono-cross {
    bottom: 2px;
    left: 2px;
}

li.complete div .icono-check {
    display: none;
}

li.todo div .icono-cross {
    display: none;
}
