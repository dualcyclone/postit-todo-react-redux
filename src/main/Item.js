import randomColor from 'randomcolor';

export default class Item {
  constructor(id, content) {
    return {
      id: id,
      content: content,
      completed: false,
      color: randomColor({ luminosity: 'light' })
    };
  }
};
