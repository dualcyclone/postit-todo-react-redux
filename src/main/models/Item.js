import randomColor from 'randomcolor'

export class Item {
  static generateKey() {
    return btoa(Date.now())
  }

  constructor(content) {
     return {
      id: Item.generateKey(),
      content: content,
      completed: false,
      color: randomColor({ luminosity: 'light' })
    }
  }
}

export default Item
