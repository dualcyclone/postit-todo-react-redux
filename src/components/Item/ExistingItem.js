import React from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { completeItem, removeItem } from '../../main/store/actions'
import Item from './Item'
import './styles.css'

export const ExistingItem = ({ item, onDoneClick, onRemoveClick }) => (
  <Item
    className={ item.completed ? 'complete' : 'todo' }
    color={ item.color }>
      <p>{ item.content }</p>
      <a onClick={onDoneClick} className="icon icono-cross">&nbsp</a>
      <a onClick={onDoneClick} className="icon icono-check">&nbsp</a>
      <a onClick={onRemoveClick} className="icon icono-trash">&nbsp</a>
  </Item>
)

ExistingItem.propTypes = {
  item: PropTypes.object.isRequired,
  onDoneClick: PropTypes.func.isRequired,
  onRemoveClick: PropTypes.func.isRequired
}

const mapDispatchToProps = (dispatch, ownProps) => ({
  onDoneClick: () => dispatch(completeItem(ownProps.item)),
  onRemoveClick: () => dispatch(removeItem(ownProps.item))
})

export default connect(
  null,
  mapDispatchToProps
)(ExistingItem)
