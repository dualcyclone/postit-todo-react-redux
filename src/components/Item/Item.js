import React from 'react';
import PropTypes from 'prop-types';
import './styles.css';

const Item = ({ ElementType = 'li', color, children, className }) => (
  <ElementType className={ 'todoItem ' + className }>
    <div style={{ backgroundColor: color }}>
      { children }
    </div>
  </ElementType>
);

Item.propTypes = {
  className: PropTypes.string.isRequired, 
  children: PropTypes.array.isRequired
};

export default Item;
