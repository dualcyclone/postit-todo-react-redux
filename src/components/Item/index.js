import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { completeItem, removeItem } from '../../main/actions';
import './styles.css';

const Item = ({ item, children, onDoneClick, onRemoveClick }) => (
  <li className={ item.completed ? 'complete' : 'todo' }>
    <div style={{ backgroundColor: item.color }}>
      <p>{ children }</p>
      <a onClick={onDoneClick} className="icon icono-cross">&nbsp;</a>
      <a onClick={onDoneClick} className="icon icono-check">&nbsp;</a>
      <a onClick={onRemoveClick} className="icon icono-trash">&nbsp;</a>
    </div>
  </li>
);

Item.propTypes = {
  item: PropTypes.object.isRequired,
  children: PropTypes.string.isRequired,
  onDoneClick: PropTypes.func.isRequired,
  onRemoveClick: PropTypes.func.isRequired
};

const mapDispatchToProps = (dispatch, ownProps) => ({
  onDoneClick: () => dispatch(completeItem(ownProps.item)),
  onRemoveClick: () => dispatch(removeItem(ownProps.item))
});

export default connect(
  null,
  mapDispatchToProps
)(Item);
