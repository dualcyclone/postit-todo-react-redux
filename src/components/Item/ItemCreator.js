import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { addItem } from '../../main/actions';
import Item from './Item';
import './styles.css';

export class ItemCreator extends Component {
  constructor(props) {
    super(props);
    this.state = {
      inputValue: ''
    };
  }

  render() {
    return (
      <Item ElementType='div' className='itemCreator'>
        <p>Add another item...</p>
        <textarea
          ref="inputField"
          className={'itemCreator-input'}
          type="text"
          rows="8"
          placeholder="What do you need to do?"
          onKeyPress={(e) => {
            if (e.key === 'Enter') {
              e.preventDefault();
              this.props.onAdd(this.state.inputValue);
              this.setState({
                inputValue: ''
              })
            }
          }}
          value={this.state.inputValue}
          onChange={evt => this.updateInputValue(evt)}
        />
        <a onClick={() => {
            this.state.inputValue && this.props.onAdd(this.state.inputValue);
            this.setState({
              inputValue: ''
            })
          }} className='icon icono-plusCircle'>&nbsp;</a>
      </Item>
    );
  }

  updateInputValue(evt) {
    this.setState({
      inputValue: evt.target.value
    });
  }
}

ItemCreator.propTypes = {
  onAdd: PropTypes.func.isRequired
};

const mapDispatchToProps = (dispatch, ownProps) => ({
  onAdd: (newItem) => dispatch(addItem(newItem))
});

export default connect(
  null,
  mapDispatchToProps
)(ItemCreator);
