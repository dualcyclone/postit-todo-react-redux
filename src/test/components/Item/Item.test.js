import React from 'react'
import { expect } from 'chai'
import { shallow, configure } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import toJson from 'enzyme-to-json'
import { Item } from '../../../../src/components/Item/Item'

const jestExpect = global.expect

configure({adapter: new Adapter()})

const defaultProps = {
  className: 'todo',
  children: [
    <p key="1">hello 1</p>,
    <p key="2">hello 2</p>
  ]
}

describe('Item', () => {
  it('Should render without failing', () => {
    const renderedItem = shallow(<Item {...defaultProps} />)

    jestExpect(toJson(renderedItem)).toMatchSnapshot()
  })

  it('Should render with the class of "todoItem"', () => {
    const renderedItem = shallow(
      <Item {...defaultProps} />
    )

    expect(renderedItem.is('.todoItem')).to.equal(true)
  })

  it('Should render with the class of "DEMO" when provided with a className as such', () => {
    const renderedItem = shallow(
      <Item {...defaultProps} className="DEMO"/>
    )

    expect(renderedItem.is('.todoItem.DEMO')).to.equal(true)
  })

  it('Should render as an "li" when no ElementType is provided', () => {
    const renderedItem = shallow(
      <Item {...defaultProps} />
    )

    expect(renderedItem.is('li')).to.equal(true)
  })

  it('Should render as a "div" when the ElementType is "div"', () => {
    const renderedItem = shallow(
      <Item {...defaultProps} ElementType='div'/>
    )

    expect(renderedItem.is('div')).to.equal(true)
  })

  it('Child div should render with a background colour when a valid colour is provided', () => {
    const expectedBackgroundColor = '#456'
    const renderedItem = shallow(
      <Item {...defaultProps} color={ expectedBackgroundColor }/>
    )

    expect(renderedItem.find('div').get(0).props.style.backgroundColor).to.equal(expectedBackgroundColor)
  })

  it('Should render provided children within its own wrapper elements', () => {
    const renderedItem = shallow(
      <Item {...defaultProps} />
    )

    expect(renderedItem.find('div').children().length).to.equal(2)
    expect(renderedItem.find('div').children().get(0)).to.eql(defaultProps.children[0])
    expect(renderedItem.find('div').children().get(1)).to.eql(defaultProps.children[1])
  })
})
