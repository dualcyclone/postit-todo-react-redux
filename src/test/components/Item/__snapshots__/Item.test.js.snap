// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Item Should render without failing 1`] = `
<li
  className="todoItem todo"
>
  <div
    style={
      Object {
        "backgroundColor": undefined,
      }
    }
  >
    <p
      key="1"
    >
      hello 1
    </p>
    <p
      key="2"
    >
      hello 2
    </p>
  </div>
</li>
`;
